File: src/org/elixir_lang/annonator/Module.java
package org.elixir_lang.annonator;

import com.intellij.lang.annotation.Annotation;
import com.intellij.lang.annotation.Annotator;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import org.elixir_lang.psi.NamedElement;
import org.elixir_lang.psi.stub.index.AllName;
import org.intellij.erlang.psi.ErlangAtom;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public class Module implements Annotator {

    public static final String ELIXIR_ALIAS_PREFIX = "Elixir.";

    @Override
    public void annotate(@NotNull PsiElement psiElement, @NotNull com.intellij.lang.annotation.AnnotationHolder annotationHolder) {
        if (psiElement instanceof ErlangAtom) {
            ErlangAtom erlangAtom = (ErlangAtom) psiElement;
            String name = erlangAtom.getName();

            if (name.startsWith(ELIXIR_ALIAS_PREFIX)) {
                Project project = psiElement.getProject();

                Collection<NamedElement> namedElementCollection = StubIndex.getElements(
                        AllName.KEY,
                        name,
                        project,
                        GlobalSearchScope.allScope(project),
                        NamedElement.class
                );

                if (namedElementCollection.size() > 0) {
                    TextRange textRange = psiElement.getTextRange();
                    String unprefixedName = name.substring(ELIXIR_ALIAS_PREFIX.length(), name.length());
                    Annotation annotation = annotationHolder.createInfoAnnotation(textRange, "Resolves to Elixir Module " + unprefixedName);
                    annotation.setTextAttributes(DefaultLanguageHighlighterColors.LINE_COMMENT);
                } else {
                    TextRange textRange = psiElement.getTextRange();
                    annotationHolder.createErrorAnnotation(textRange, "Unresolved Elixir Module");
                }
            }
        }
    }
}

File: src/org/elixir_lang/annotator/Module.kt
package org.elixir_lang.annotator

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.psi.PsiElement
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import org.elixir_lang.psi.NamedElement
import org.elixir_lang.psi.stub.index.ModularName
import org.intellij.erlang.psi.ErlangAtom

class Module : Annotator {
    override fun annotate(psiElement: PsiElement, annotationHolder: AnnotationHolder) {
        if (psiElement is ErlangAtom) {
            val name = psiElement.name
            if (name.startsWith(ELIXIR_ALIAS_PREFIX)) {
                val project = psiElement.getProject()
                val namedElementCollection = StubIndex.getElements(
                        ModularName.KEY,
                        name,
                        project,
                        GlobalSearchScope.allScope(project),
                        NamedElement::class.java
                )
                if (namedElementCollection.size > 0) {
                    val unprefixedName = name.substring(ELIXIR_ALIAS_PREFIX.length)
                    annotationHolder
                            .newAnnotation(HighlightSeverity.WEAK_WARNING, "Resolves to Elixir Module $unprefixedName")
                            .textAttributes(DefaultLanguageHighlighterColors.LINE_COMMENT)
                            .create()
                } else {
                    annotationHolder
                            .newAnnotation(HighlightSeverity.ERROR, "Unresolved Elixir Module")
                            .create()
                }
            }
        }
    }

    companion object {
        const val ELIXIR_ALIAS_PREFIX = "Elixir."
    }
}
